NCGenerics: synthesize Copyable/Escapable decls

kavon opened the pull request.

kavon committed:
[NCGenerics] test for Mutator pattern

covers rdar://117867086

kavon committed:
[NCGenerics] add basic end-to-end test

eeckstein commented on include/swift/SIL/SILType.h:
```diff
@@ -763,9 +763,13 @@ class SILType {
+  ///
+  /// orWrapped == false -->  isNoncopyable
+  /// orWrapped == true  -->  isNoncopyable || isMoveOnlyWrapped
+  bool isMoveOnly(bool orWrapped=true) const;
```
At this occasion, can you rename this API to `isCopyable` (with the invert meaning)?

kavon replied:
Something that's wrapped in a SILMoveOnlyWrapper isn't necessarily noncopyable. It's just not implicitly copyable, so perhaps it should really imply that instead.

eeckstein reviewed (commented):
Does this actually change something in the mangling scheme which requires changes in the demangler?

kavon commented:
Mangler changes are needed to mangle the uses of `~Copyable` into generic signatures, but that's now moving to another PR for now (Sorry for changing the focus of this one).

kavon commented:
@swift-ci please test

kavon committed:
NCGenerics: improve assertion messages

kavon committed:
Test: prefer `-enable-builtin-module`

Many tests that use `-parse-stdlib` only do so because they want access
to the Builtin module. Now that we have the flag and ability to import
it, use that instead.

kavon commented:
@swift-ci please test

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci test

kavon committed:
NCGenerics: remove unreachable and explain

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci smoke test

ahoppen reviewed (commented):
Parser changes look good to me. It probably needs some adjustment for code completion though. It would also be good to change the `complete_wihout_constraint_type.swift` test to check that we only suggest `Copyable` after `~` and not all the types by changing it to the following (which currently passes).

```swift
// RUN: %batch-code-completion

struct FileDescriptor: ~#^COMPLETE^# {}
// COMPLETE: Begin completions, 1 item
// COMPLETE-DAG: Decl[Protocol]/OtherModule[Swift]/IsSystem: Copyable[#Copyable#]; name=Copyable
```

kavon commented:
@swift-ci smoke test

kavon committed:
NCGenerics: sometimes synth. Copyable/Escapable

When the Swift module is not available, we'll synthesize the
Copyable/Escapable decls into the Builtin module.

In the future, it might be nice to just do this always, and define
typealiases for those types in the stdlib to refer to the ones in the
builtin module.

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci smoke test

kavon committed:
NCGenerics: update a number of tests

These tests were not updated in the transition to
`REQUIRES: noncopyable_generics` and thus running them with a
correctly-built stdlib that has the Copyable requirements.

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci smoke test

slavapestov commented on lib/Sema/TypeCheckInvertible.cpp:
```diff
@@ -468,13 +468,9 @@ ProtocolConformance *deriveConformanceForInvertible(Evaluator &evaluator,
-    }
-
+    // There are various other kinds of SourceFiles, like SIL, which instead
+    // get their conformances here instead.
```
Why don't source file nominals also get their conformance from here then?
